aptos-move/dynamic-transaction-composer/src/builder.rs

```diff
@@ -0,0 +1,593 @@
+        let expected_args_ty = {
+            let script = self.builder.as_script();
+            let func = script.function_handle_at(call_idx);
+            if script.signature_at(func.parameters).0.len() != args.len() {
+                bail!(
+                    "Function {}::{} argument call size mismatch",
+                    module,
+                    function
+                );
+            }
+            script
+                .signature_at(func.parameters)
+                .0
+                .iter()
+                .map(|ty| ty.instantiate(&subst_mapping))
+                .collect::<Vec<_>>()
+        };
+
+        for (arg, ty) in args.into_iter().zip(expected_args_ty) {
+            match arg {
+                Argument::Local(i) => {
+                    self.check_argument_compatibility(&i, &ty)?;
+                    arguments.push(i)
+                },
+                Argument::Signer(i) => arguments.push(AllocatedLocal {
+                    op_type: ArgumentOperation::Copy,
+                    is_parameter: true,
+                    local_idx: i,
+                }),
+                Argument::Raw(bytes) => {
+                    let new_local_idx = self.parameters_ty.len() as u16;
+                    self.parameters_ty.push(ty);
+                    self.parameters.push(bytes);
+                    arguments.push(AllocatedLocal {
+                        op_type: ArgumentOperation::Move,
+                        is_parameter: true,
+                        local_idx: new_local_idx,
+                    })
+                },
+            }
+        }
+
+        let mut returns = vec![];
+
+        let expected_returns_ty = {
+            let func = self.builder.as_script().function_handle_at(call_idx);
+            self.builder
+                .as_script()
+                .signature_at(func.return_)
+                .0
+                .iter()
+                .map(|ty| ty.instantiate(&subst_mapping))
+                .collect::<Vec<_>>()
+        };
+
+        for return_ty in expected_returns_ty {
+            let local_idx = self.locals_ty.len() as u16;
+            self.locals_ty.push(return_ty);
+            self.locals_availability.push(true);
+            returns.push(local_idx);
+        }
+
+        self.calls.push(BuilderCall {
+            type_args: call_type_params,
+            call_idx,
+            arguments,
+            returns: returns.clone(),
+            #[cfg(test)]
+            type_tags: ty_args,
+        });
+        Ok(returns
+            .into_iter()
+            .map(|idx| {
+                Argument::Local(AllocatedLocal {
+                    op_type: ArgumentOperation::Move,
+                    is_parameter: false,
+                    local_idx: idx,
+                })
+            })
+            .collect())
+    }
+
+    fn check_drop_at_end(&self) -> anyhow::Result<()> {
+        let view = BinaryIndexedView::Script(self.builder.as_script());
```

Comment:
Hmm... kinda funny that we are converting this into a view knowing it's a script.

Is it because you want to use some convenient functions from the view?

Reply:
Yea unfortunately the ability of a signature token is only implemented with view.